pennylane/transforms/compile.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Code for the high-level quantum function transform that executes compilation."""
+
+
+from pennylane import apply
+from pennylane.tape import get_active_tape
+from pennylane.ops import __all__ as all_ops
+
+from pennylane.transforms import qfunc_transform
+from pennylane.transforms.optimization import cancel_inverses, commute_controlled, merge_rotations
+
+
+default_pipeline = [commute_controlled, cancel_inverses, merge_rotations]
+
+
+@qfunc_transform
+def compile(tape, pipeline=None, basis_set=None, num_passes=1):
+    """Compile a circuit by applying a series of transforms to a quantum function.
+
+    The default set of transforms includes (in order):
+
+    - pushing all commuting single-qubit gates as far right as possible
+      (:func:`~.commute_controlled`)
+    - cancellation of adjacent inverse gates
+      (:func:`~.cancel_inverses`)
+    - merging adjacent rotations of the same type
+      (:func:`~.merge_rotations`)
+
+    Args:
+        qfunc (function): A quantum function.
+        pipeline (list[qfunc_transform]): A list of quantum function transforms
+            to apply.
+        basis_set (list[str]): A list of basis gates. When expanding the tape,
+            expansion will continue until gates in the specific set are
+            reached. If no basis set is specified, no expansion will be done.
+
+        num_passes (int): The number of times to apply the set of transforms in
+            ``pipeline``. The default is to perform each transform once;
+            however, doing so may produce a new circuit where applying the set
+            of transforms again may yield further improvement, so the number of
+            such passes can be adjusted.
```

Comment:
would there ever be instances where it would be useful  to apply each transform a different number of times? Could this be specified by passing a `list` of `ints`?

Reply:
Yup, but I think it is more straightforward to repeat the transform in the pipeline, e.g.,:
```
pipeline = [commute_controlled, cancel_inverses, commute_controlled]
```
Then `num_passes` is reserved for full applications of the pipeline.